packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/decode/DecodeForm.tsx

```diff
@@ -40,30 +63,113 @@ function isAbiError(item: AbiItem): item is AbiItem & { type: "error" } {
+  const [decoded, setDecoded] = useState<AbiFunction | AbiError | ResourceItem | SelectorDatabase>();
+  const [encoded, setEncoded] = useState<string | undefined>();
   const form = useForm<z.infer<typeof formSchema>>({
     resolver: zodResolver(formSchema),
   });
 
-  function onSubmit({ selector }: z.infer<typeof formSchema>) {
+  const onSubmit = async ({ encodedData }: z.infer<typeof formSchema>) => {
+    setEncoded(encodedData);
+    if (!encodedData) return;
+
+    const selector = encodedData.substring(0, 10);
     const worldAbi = worldData?.abi || [];
     const systemsAbis = systemData ? Object.values(systemData) : [];
     const abis = [worldAbi, ...systemsAbis].flat();
 
+    // Try to find matching ABI item
     const abiItem = abis.find((item): item is AbiFunction | AbiError => {
       if (isAbiFunction(item)) {
         return toFunctionSelector(item) === selector;
-      } else if (isAbiError(item)) {
+      }
+      if (isAbiError(item)) {
         return getErrorSelector(item) === selector;
       }
       return false;
     });
 
-    setAbiItem(abiItem);
-  }
+    if (abiItem) {
+      setDecoded(abiItem);
+      return;
+    }
+
+    // Try to decode as resource
+    try {
+      const resource = hexToResource(encodedData as Hex);
+      if (resource) {
+        setDecoded({ type: "resource", id: encodedData, resource });
+        return;
+      }
+    } catch (error) {
+      console.error("Error decoding resource:", error);
+    }
+
+    // Try to find in 4-byte database
+    if (selector.length !== 10) return;
+    try {
+      const response = await fetch(`https://www.4byte.directory/api/v1/signatures/?hex_signature=${selector}`);
+      const data = await response.json();
+      if (data.results.length > 0) {
+        setDecoded({
+          type: "signature",
+          selectors: data.results.map((result: { text_signature: string }) => result.text_signature),
+        });
+      }
+    } catch (error) {
+      console.error("Error fetching 4byte data:", error);
+    }
+  };
+
+  const results = useMemo<Result[]>(() => {
+    if (!decoded || !encoded) return [];
+    try {
+      if (decoded.type === "resource") {
+        return [{ type: decoded.resource.type, label: resourceToLabel(decoded.resource) }];
+      }
+      if (decoded.type === "function" || decoded.type === "error") {
+        const decodedCall = decodeAbiItem(decoded, encoded as Hex);
+        return [{ type: decoded.type, label: formatAbiItem(decoded), decodedCall }];
+      }
+      if (decoded.type === "signature") {
+        return decoded.selectors.map((selector) => {
+          const abiItem = parseAbiItem(`function ${selector}`) as AbiFunction;
+          const decodedCall = decodeAbiItem(abiItem, encoded as Hex);
+          return { type: "signature", label: selector, decodedCall };
+        });
+      }
+    } catch (error) {
+      console.error("Error decoding function data:", error);
+    }
```

Comment:
this now throws when attempting to decode a selector without calldata (ie `0x170a8db6`) - think we should wrap `decodeAbiItem` in another try/catch so it just doesn't show the params if decoding the call failed but decoding the selector succeeded